### DISCUSSION TITLE
Integrate a "scheduler" to control scheduled jobs within your codebase

### DISCUSSION THREAD

**Discussion by aaronbushnell** [craftcms/cms#8847] [Ideas]
### Description
[Laravel has a "task scheduling"](https://laravel.com/docs/master/scheduling) feature that allows you to handle your recurring jobs directly in your codebase. For instance:

```sh
# Checks every minute for scheduled tasks
* * * * * cd /path-to-your-project && php artisan schedule:run >> /dev/null 2>&1
```

```php
$schedule->job(new Heartbeat)->everyFiveMinutes();

// Dispatch the job to the "heartbeats" queue...
$schedule->job(new Heartbeat, 'heartbeats')->everyFiveMinutes();

// Run hourly from 8 AM to 5 PM on weekdays...
$schedule->command('foo')
          ->weekdays()
          ->hourly()
          ->timezone('America/Chicago')
          ->between('8:00', '17:00');
```

There's [a Yii2 port](https://github.com/omnilight/yii2-scheduling) of Laravel's Schedule component, however it's a bit tricky to implement this in Craft because it is designed to run using:

```sh
php /path/to/yii yii schedule/run
```

Including this feature in Craft core would be extremely useful for projects that have numerous cron jobs. This would allow a developer to "check in" their scheduled tasks as part of the codebase, version them, and easily port them when switching from server to server without recreating numerous cron jobs during the migration process.

### Steps to reproduce

N/A

### Additional info

- Craft version:  Craft Pro 3.4.24
- PHP version: 7.2.23
- Database driver & version: MySQL 5.7.29
- Plugins & versions: None

**Comment by brandonkelly** [comment]
We’re using that exact library [internally](https://github.com/pixelandtonic/craftnet/blob/master/config/schedule.php), and it works great. Definitely considering making scheduling a baked-in feature of Craft, as there are a few tasks that would be better off done on a schedule such as garbage collection, publishing/expiring entries, etc., but that will require setting up a cron job, so not something we can do until v4.

**Comment by aaronbushnell** [comment]
Really cool, @brandonkelly! Looking forward to seeing something like this in core. Understandable that it requires a major version update for it. Thanks!

**Comment by brandonkelly** [comment]
Related: #978

**Comment by BenParizek** [comment]
I have several plugin use cases for this (report generation, transactional email) and would love to piggy back on a standardized Craft way of doing this to make it easy for developers to manage all of this stuff in once place.
